File: trunk/xapian-bindings/python/docs/examples/simplematchdecider.py
#!/usr/bin/env python
#
# $Id$
# Simple command-line match decider example
#
# ----START-LICENCE----
# Copyright 2003 James Aylett
# Copyright 2004 Olly Betts
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA
# -----END-LICENCE-----

import sys
import xapian

if len(sys.argv) < 4:
    print >> sys.stderr, "usage: %s <path to database> <avoid value> <search terms>" % sys.argv[0]
    sys.exit(1)

class mymatcher(xapian.MatchDecider):
    def __init__(self, avoidvalue):
        xapian.MatchDecider.__init__(self)
        self.avoidvalue = avoidvalue
        
    def __call__(self, doc):
        return doc.get_value(0) != self.avoidvalue

try:
    database = xapian.Database(sys.argv[1])

    enquire = xapian.Enquire(database)
    stemmer = xapian.Stem("english")
    terms = []
    for term in sys.argv[3:]:
        terms.append(stemmer.stem_word(term.lower()))
    query = xapian.Query(xapian.Query.OP_OR, terms)
    print "Performing query `%s'" % query.get_description()

    enquire.set_query(query)
    matcher = mymatcher(sys.argv[2])
    matches = enquire.get_mset(0, 10, None, matcher)

    print "%i results found" % matches.get_matches_estimated()
    for match in matches:
        print "ID %i %i%% [%s]" % (match[xapian.MSET_DID], match[xapian.MSET_PERCENT], match[xapian.MSET_DOCUMENT].get_data())

except Exception, e:
    print >> sys.stderr, "Exception: %s" % str(e)
    sys.exit(1)

File: trunk/xapian-bindings/lua/docs/examples/simplematchdecider.lua
#!/usr/bin/env lua
--
-- Simple command-line match decider example.
--
-- Copyright (C) 2011 Xiaona Han
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License as
-- published by the Free Software Foundation; either version 2 of the
-- License, or (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301
-- USA

require("xapian")

-- This example runs a query like simplesearch does, but uses a MatchDecider
-- (mymatchdecider) to discard any document for which value 0 is equal to
-- the string passed as the second command line argument.

if #arg < 3 then
	io.stderr:write("Usage:" .. arg[0] .. " PATH_TO_DATABASE AVOID_ID QUERY\n")
	os.exit()
end

-- My own matchdecider
function mymatchdecider(doc)
	return doc:get_docid() ~= avoid_id
end

-- Open the database for searching.
database = xapian.Database(arg[1])

-- Start an enquire session.
enquire = xapian.Enquire(database)

-- Save the second command line argument into the variable and combine the rest
-- of the command line arguments with spaces between them, so that simple
-- queries don't have to be quoted at the shell level.
avoid_id = tonumber(arg[2])
query_string = table.concat(arg, " ", 3, #arg)

-- Parse the query string to produce a Xapian::Query object.
qp = xapian.QueryParser()
stemmer = xapian.Stem("english")
qp:set_stemmer(stemmer)
qp:set_database(database)
qp:set_stemming_strategy(xapian.QueryParser_STEM_SOME)
query = qp:parse_query(query_string)
print("Parsed query is: " .. tostring(query))

-- Find the top 10 results for the query.
enquire:set_query(query)
matches = enquire:get_mset(0, 10, None, mymatchdecider)

-- Display the size of the results.
print(string.format("%u results found.", matches:get_matches_estimated()))
print(string.format("Results 1-%i:", matches:size()))

-- Display the results
for m in matches:items() do
	print(m:get_rank() + 1, m:get_percent() .. "%", m:get_docid(), m:get_document():get_data())
end
